featurebyte/query_graph/sql/ast/input.py

```diff
@@ -162,6 +167,63 @@ def make_input_columns_map(cls, context: SQLNodeContext) -> dict[str, Expression
         """
         columns_map: dict[str, Expression] = {}
         for col_dict in context.parameters["columns"]:
-            colname = col_dict["name"]
-            columns_map[colname] = expressions.Identifier(this=colname, quoted=True)
+            column_spec = ColumnSpec(**col_dict)
+            colname = column_spec.name
+            timestamp_schema = column_spec.timestamp_schema
+            if (
+                timestamp_schema is not None
+                and context.parameters["type"] == TableDataType.SCD_TABLE
+            ):
+                column_expr = cls._get_generic_timestamp_column_expr(
+                    column_spec, timestamp_schema, context.adapter
+                )
+            else:
+                column_expr = expressions.Identifier(this=colname, quoted=True)
+            columns_map[colname] = column_expr
         return columns_map
+
+    @classmethod
+    def _get_generic_timestamp_column_expr(
+        cls,
+        column_spec: ColumnSpec,
+        timestamp_schema: TimestampSchema,
+        adapter: BaseAdapter,
+    ) -> Expression:
+        """
+        Get expression for a timestamp column converted to UTC
+
+        Parameters
+        ----------
+        column_spec : ColumnSpec
+            Column specification
+        timestamp_schema : TimestampSchema
+            Timestamp schema
+        adapter : BaseAdapter
+            SQL adapter
+
+        Returns
+        -------
+        Expression
+        """
+        # Original datetime column (could be a timestamp, date or string)
+        column_expr: Expression = expressions.Identifier(this=column_spec.name, quoted=True)
+
+        if column_spec.dtype == DBVarType.VARCHAR:
+            assert timestamp_schema.format_string is not None
+            column_expr = adapter.to_timestamp_from_string(
+                column_expr, timestamp_schema.format_string
+            )
+        elif column_spec.dtype == DBVarType.DATE:
+            # Treat date type columns as end of day in the local timezone
+            column_expr = adapter.dateadd_second(make_literal_value(86400), column_expr)
```

Comment:
Should this be `86399`? If we use strict less than for cutoffs then this should be ok

Reply:
For SCD joins we use `<=` condition for cutoffs. 

For a row in an SCD table with an effective date value of `2024-01-15`, what is the earliest point in time that can be used to lookup that row?

1. `2024-01-16 00:00:00`
2. `2024-01-15 23:59:59`

The current version implements 1). Using `86399` would be 2). I thought 1) is cleaner and more conservative. Either way, only the point-in-times within that one second are affected. Let me know if you have a preference and if I should change it.

Reply:
I think we can stick with option 1